README.md
# pyetc




## python ETC module

## Description
pyetc is a python ETC module developed for the WST project



docs/changelog.rst
Changelog
=========

0.30 (2/12/2023)
----------------
- use different obscuration for the MOS and the IFS
- add redshifting of the template
- use MOS-LR with 3 channels
- correct the MOS-LR throughputs using new values provided by Roelof & Jarle
- updated tutorial to refelect the change

0.22 (14/10/2023)
-----------------
- Correct ima.fwhm_gauss for older version of mpdaf
- remove np.float (deprecated in new numpy version)

0.21 (18/02/2023)
-----------------
- Correct telescope aperture to take into account 20% central obscuration

0.2 (16/02/2023)
-----------------
- First release to the consortium
- Include provisional model for IFS, MOS-LR and MOS-HR

New features
^^^^^^^^^^^^
- improved documentation (docstrings)
- documentation available on readthedocs
- sersic model in get_ima


0.1 (25/01/2023)
-----------------

First beta version


tests/test_ps.py
# Test point source SN computation
from pyetc.wst import WST
from numpy.testing import assert_almost_equal
from mpdaf.obj import flux2mag

etc = WST(log='INFO')

ifs = etc.ifs['blue']
obs = dict(
    ndit = 2, 
    dit = 1800, 
    nb_spectels = 3,
    airmass = 1,
    seeing = 0.7,
    frac_flux = 0.8,
)
etc.set_obs(obs)

frac_flux,nspaxels,rad   = etc.fluxfrac(ifs)
assert (frac_flux > 0) and (frac_flux < 1)
assert nspaxels > 1
assert (rad > 0) and (rad < 2.0)

flux = 1e-18
res = etc.snr_from_psflux(ifs, flux, 'darksky')
assert_almost_equal(res['flux'], flux)
assert res['snr'].data.min() > 0
assert res['snr'].data.max() > res['snr'].data.min()
assert res['snr'].shape[0] == ifs['instrans'].shape[0]
assert res['noise']['tot'].shape[0] == ifs['instrans'].shape[0]

k = 2000
wave = res['snr'].wave.coord(k)
snr0 = res['snr'].data[k]

res3 = etc.psflux_from_snr(ifs, snr0, 'darksky')
assert_almost_equal(res3['flux'].data[k]*1.e18, flux*1.e18)

mag = flux2mag(flux/ifs['dlbda'], 0, wave)[0]
res2 = etc.snr_from_psmag(ifs, mag, 'darksky')
assert_almost_equal(res2['snr'].data[k], snr0)

res4 = etc.psmag_from_snr(ifs, snr0, 'darksky')
assert_almost_equal(res4['mag'].data[k], mag)

mos = etc.moslr['red']
frac_flux,nbspaxels,rad = etc.fluxfrac(mos)
assert (frac_flux > 0) and (frac_flux < 1)

flux = 1e-18
res = etc.snr_from_psflux(mos, flux, 'darksky')
assert_almost_equal(res['flux'], flux)
assert res['snr'].data.min() > 0
assert res['snr'].data.max() > res['snr'].data.min()
assert res['snr'].shape[0] == mos['instrans'].shape[0]
assert res['noise']['tot'].shape[0] == mos['instrans'].shape[0]

k = 2000
wave = res['snr'].wave.coord(k)
mag = flux2mag(flux/mos['dlbda'], 0, wave)[0]
snr0 = res['snr'].data[k]

res2 = etc.snr_from_psmag(mos, mag, 'darksky')
assert_almost_equal(res2['snr'].data[k], res['snr'].data[k])

res3 = etc.psflux_from_snr(mos, snr0, 'darksky')
assert_almost_equal(res3['flux'].data[k]*1.e18, flux*1.e18)

res4 = etc.psmag_from_snr(mos, snr0, 'darksky')
assert_almost_equal(res4['mag'].data[k], mag)


tests/test_sb.py
# Test surface brightness SN computation
from pyetc.wst import WST
from numpy.testing import assert_almost_equal
from mpdaf.obj import flux2mag

etc = WST(log='INFO')

ifs = etc.ifs['blue']
obs = dict(
    ndit = 2, 
    dit = 1800,
    nb_spaxels = 4 * 4, 
    nb_spectels = 3,
    airmass = 1,
)
etc.set_obs(obs)
sbflux = 1e-18 
res = etc.snr_from_sb(ifs, sbflux, 'darksky')
assert res['area'] == 1.0, 'area != 1.0'
assert_almost_equal(res['sbflux'], sbflux)
assert res['snr'].data.min() > 0
assert res['snr'].data.max() > res['snr'].data.min()
assert res['snr'].shape[0] == ifs['instrans'].shape[0]
assert res['noise']['tot'].shape[0] == ifs['instrans'].shape[0]

k = 2000
wave = res['snr'].wave.coord(k)
mag = flux2mag(sbflux, 0, wave)[0]
snr0 = res['snr'].data[k]
res2 = etc.snr_from_esmag(ifs, mag, 'darksky')
assert_almost_equal(res2['snr'].data[k], res['snr'].data[k])

res3 = etc.sb_from_snr(ifs, snr0, 'darksky')
assert_almost_equal(res3['sbflux'].data[k]*1.e18, sbflux*1.e18, decimal=3)

res4 = etc.esmag_from_snr(ifs, snr0, 'darksky')
assert_almost_equal(res4['mag'].data[k], mag)


tests/testmain.py
# Test main

from pyetc.etc import vdisp2sigma
from pyetc.wst import WST
from mpdaf.obj import Spectrum, WaveCoord, mag2flux, flux2mag
import numpy as np
from matplotlib import pyplot as plt



print('start')
wst = WST(log='DEBUG')

mosr = wst.moslr['red']

obs = dict(
    moon = 'darksky',
    airmass = 1,
    seeing = 0.7,
    ndit = 2,
    dit = 1800, 
    spec_type = 'cont',
    ima_type = 'ps',
)
wst.set_obs(obs)

snr = 3 #by A
wave = 8000
dspec = dict(type='flatcont', wave=[wave-10,wave+10])
spec = wst.get_spec(mosr, dspec)
res = wst.flux_from_source(mosr, snr, None, spec)
print('end')


pyetc/vlt.py
import logging
from pyetc.version import  __version__
from mpdaf.obj import Spectrum, WaveCoord
from astropy.table import Table
import os, sys
import numpy as np
from mpdaf.log import setup_logging
import astropy.units as u
from .etc import ETC, get_data


CURDIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'data/VLT')

class VLT(ETC):
    
    def __init__(self, log=logging.INFO, skip_dataload=False):
        self.refdir = CURDIR
        self.version = __version__
        self.logger = logging.getLogger(__name__)
        setup_logging(__name__, level=log, stream=sys.stdout)        
        # Telescope
        self.name = 'VLT'
        self.tel = dict(area = 48.5,# effective aperture in squared meter 
                        diameter = 8.20, # pirmary diameter in meter
                        )      
        # IFS
        self.ifs = {} # 
        self.ifs['channels'] = ['blue','red']
        
        # IFS BlueMUSE
        chan = 'blue'
        self.ifs[chan] = dict(desc = 'BlueMUSE 5/01/2022',
                                type='IFS',
                                version='1.0', 
                                iq_fwhm = 0.10, # fwhm PSF of telescope + instrument
                                iq_beta = 2.50, # beta PSF of telescope + instrument                                
                                spaxel_size = 0.20, # spaxel size in arcsec
                                dlbda = 0.575, # Angstroem/pixel
                                lbda1 = 3500, # starting wavelength in Angstroem
                                lbda2 = 5800, # end wavelength in Angstroem
                                lsfpix = 2.0, # LSF in spectel
                                ron = 3.0, # readout noise (e-)
                                dcurrent = 3.0, # dark current (e-/pixel/h)                                
                                )
        
        if not skip_dataload:
            get_data(self.ifs, chan, 'ifs', CURDIR)
        
        # IFS MUSE
        chan = 'red'
        self.ifs[chan] = dict(desc = 'MUSE 5/01/2022',
                               type='IFS',
                               version='1.0', 
                               iq_fwhm = 0.10, # fwhm PSF of telescope + instrument
                               iq_beta = 2.50, # beta PSF of telescope + instrument                               
                               spaxel_size = 0.20, # spaxel size in arcsec
                               dlbda = 1.25, # Angstroem/pixel
                               lbda1 = 4800, # starting wavelength in Angstroem
                               lbda2 = 9300, # end wavelength in Angstroem
                               lsfpix = 2.5, # LSF in spectel
                               ron = 3.0, # readout noise (e-)
                               dcurrent = 3.0, # dark current (e-/pixel/h)
                               )
        if not skip_dataload:
            get_data(self.ifs, chan, 'ifs', CURDIR)   
        
        # MOS Giraffe Medusa
        self.giraffe = {} 
        self.giraffe['channels'] = ['blue']       
        # MOS-LR blue channel 
        chan = self.giraffe['channels'][0]
        self.giraffe[chan] = dict(desc='Based on ESO ETC 17/01/2023',
                                  version='1.0', 
                                  type = 'MOS',
                                  iq_fwhm = 0.40, # fwhm PSF of telescope + instrument
                                  iq_beta = 2.50, # beta PSF of telescope + instrument                                  
                                  spaxel_size = 0.30, # spaxel size in arcsec
                                  aperture = 1.8, # fiber diameter in arcsec
                                  dlbda = 0.12, # Angstroem/pixel
                                  lbda1 = 3702.69, # starting wavelength in Angstroem
                                  lbda2 = 4080.45, # end wavelength in Angstroem
                                  lsfpix = 4.0, # LSF in spectel
                                  ron = 4.0, # readout noise (e-)
                                  dcurrent = 0.5, # dark current (e-/pixel/h)                                
                                ) 
        if not skip_dataload:
            get_data(self.giraffe, 'blue', 'giraffe', CURDIR)                    
        
    def info(self):
        self._info(['ifs', 'giraffe'])        
        


pyetc/wst.py
import logging
from pyetc.version import  __version__
from mpdaf.obj import Spectrum, WaveCoord
from astropy.table import Table
import os, sys
import numpy as np
from mpdaf.log import setup_logging
import astropy.units as u
from .etc import ETC, get_data

CURDIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'data/WST')

# Cass design BusyWeek Oct 2023
MOS_OBSCURATION = 0.10 # telescope obscuration in the telescope MOS path
IFS_OBSCURATION = 0.18 # telescope obscuration in the telescope IFS path

class WST(ETC):
    
    def __init__(self, log=logging.INFO, skip_dataload=False):
        self.refdir = CURDIR
        self.version = __version__
        setup_logging(__name__, level=log, stream=sys.stdout)
        self.logger = logging.getLogger(__name__)
        self.logger.propagate = False
        
        # ------ Telescope ---------
        self.name = 'WST'
        self.tel = dict(area=100.0,  # squared meter of primary mirror (without obscuration)
                        diameter=12.0, # primary diameter
                        desc='Cass design',
                        version='Busyweek Oct 2023'
                        )        
        # ------- IFS -----------
        self.ifs = {} 
        self.ifs['channels'] = ['blue', 'red']
        # IFS blue channel
        chan = 'blue'
        self.ifs[chan] = dict(desc = 'Inspired from BlueMUSE throughput',
                              version = '0.1 10/02/2023',
                              type = 'IFS',
                              obscuration=IFS_OBSCURATION, # IFS telescope obscuration,
                              iq_fwhm = 0.10, # fwhm PSF of telescope + instrument
                              iq_beta = 2.50, # beta PSF of telescope + instrument
                              spaxel_size = 0.25, # spaxel size in arcsec
                              dlbda = 0.64, # Angstroem/pixel
                              lbda1 = 3700, # starting wavelength in Angstroem
                              lbda2 = 6100, # end wavelength in Angstroem
                              lsfpix = 2.5, # LSF in spectel
                              ron = 3.0, # readout noise (e-)
                              dcurrent = 3.0, # dark current (e-/pixel/h)                                
                              )
        if not skip_dataload:
            get_data(self.ifs, chan, 'ifs', CURDIR)                

        # IFS red channel
        chan = 'red'
        self.ifs[chan] = dict(desc='Inspired from MUSE throughput', 
                               version = '0.1 10/02/2023',
                               type='IFS',
                               obscuration=IFS_OBSCURATION, # IFS telescope obscuration,
                               iq_fwhm = 0.10, # fwhm PSF of telescope + instrument
                               iq_beta = 2.50, # beta PSF of telescope + instrument
                               spaxel_size = 0.25, # spaxel size in arcsec
                               dlbda = 0.97, # Angstroem/pixel
                               lbda1 = 6000, # starting wavelength in Angstroem
                               lbda2 = 9600, # end wavelength in Angstroem
                               lsfpix = 2.5, # LSF in spectel
                               ron = 3.0, # readout noise (e-)
                               dcurrent = 3.0, # dark current (e-/pixel/h)
                               )
        if not skip_dataload:
            get_data(self.ifs, chan, 'ifs', CURDIR)
              
        # # --------- MOSLR-VIS 3 channels 6k CCD -------------
        self.moslr = {} 
        self.moslr['channels'] = ['blue','orange', 'red']       
        # MOS-LR blue channel 
        chan = self.moslr['channels'][0]
        self.moslr[chan] = dict(desc='Inspired from 4MOST LR throughput',
                                version = '0.2 30/11/2023',
                                ref = 'throughput version revised by Jarle & Roelof',
                                type = 'MOS',
                                obscuration=MOS_OBSCURATION, # MOS telescope obscuration,
                                iq_fwhm = 0.30, # fwhm PSF of telescope + instrument
                                iq_beta = 2.50, # beta PSF of telescope + instrument
                                spaxel_size = 0.25, # spaxel size in arcsec
                                aperture = 1.0, # fiber diameter in arcsec
                                dlbda = 0.41, # Angstroem/pixel
                                lbda1 = 3700, # starting wavelength in Angstroem - we hope to go to 3500
                                lbda2 = 5350, # end wavelength in Angstroem
                                lsfpix = 4.1, # LSF in spectel
                                ron = 3.0, # readout noise (e-)
                                dcurrent = 3.0, # dark current (e-/pixel/h)                                
                                )
        if not skip_dataload:
            get_data(self.moslr, chan, 'moslr', CURDIR)
            
        # MOS-LR orange channel      
        chan = self.moslr['channels'][1] 
        self.moslr[chan] = dict(desc='Inspired from 4MOST LR throughput',
                                version = '0.2 30/11/2023',
                                ref = 'throughput version revised by Jarle & Roelof',
                                type = 'MOS',
                                obscuration=MOS_OBSCURATION, # MOS telescope obscuration,
                                iq_fwhm = 0.30, # fwhm PSF of telescope + instrument
                                iq_beta = 2.50, # beta PSF of telescope + instrument
                                spaxel_size = 0.25, # spaxel size in arcsec
                                aperture = 1.0, # fiber diameter in arcsec
                                dlbda = 0.55, # Angstroem/pixel
                                lbda1 = 5150, # starting wavelength in Angstroem
                                lbda2 = 7400, # end wavelength in Angstroem
                                lsfpix = 4.1, # LSF in spectel
                                ron = 3.0, # readout noise (e-)
                                dcurrent = 3.0, # dark current (e-/pixel/h)                                
                                )
        if not skip_dataload:
            get_data(self.moslr, chan, 'moslr', CURDIR)  


        # MOS-LR red channel      
        chan = self.moslr['channels'][2] 
        self.moslr[chan] = dict(desc='Inspired from 4MOST LR throughput',
                                version = '0.2 30/11/2023',
                                ref = 'throughput version revised by Jarle & Roelof',
                                type = 'MOS',
                                obscuration=MOS_OBSCURATION, # MOS telescope obscuration,
                                iq_fwhm = 0.30, # fwhm PSF of telescope + instrument
                                iq_beta = 2.50, # beta PSF of telescope + instrument
                                spaxel_size = 0.25, # spaxel size in arcsec
                                aperture = 1.0, # fiber diameter in arcsec
                                dlbda = 0.61, # Angstroem/pixel
                                lbda1 = 7200, # starting wavelength in Angstroem
                                lbda2 = 9700, # end wavelength in Angstroem
                                lsfpix = 4.1, # LSF in spectel
                                ron = 3.0, # readout noise (e-)
                                dcurrent = 3.0, # dark current (e-/pixel/h)                                
                                )
        if not skip_dataload:
            get_data(self.moslr, chan, 'moslr', CURDIR)  

            
        # --------- MOS-HR 4 channels  -------------
        self.moshr = {} 
        self.moshr['channels'] = ['U','B','V','I']       
        # MOS-HR U channel 
        chan = self.moshr['channels'][0]
        self.moshr[chan] = dict(desc='WST HR spectrometer possible baseline description',  
                                version = '1.1 13/02/2023',
                                ref = '',
                                type = 'MOS',
                                iq_fwhm = 0.30, # fwhm PSF of telescope + instrument
                                iq_beta = 2.50, # beta PSF of telescope + instrument
                                spaxel_size = 0.217, # spaxel size in arcsec
                                aperture = 0.93, # fiber diameter in arcsec
                                dlbda = 0.0217, # Angstroem/pixel
                                lbda1 = 3800, # starting wavelength in Angstroem
                                lbda2 = 4000, # end wavelength in Angstroem
                                lsfpix = 4.5, # LSF in spectel
                                ron = 3.0, # readout noise (e-)
                                dcurrent = 3.0, # dark current (e-/pixel/h)                                
                                )
        if not skip_dataload:
            get_data(self.moshr, chan, 'moshr', CURDIR)  
            
        # MOS-HR B channel 
        chan = self.moshr['channels'][1]
        self.moshr[chan] = dict(desc='WST HR spectrometer possible baseline description',  
                                version = '1.1 13/02/2023',
                                ref = '',
                                type = 'MOS',
                                iq_fwhm = 0.30, # fwhm PSF of telescope + instrument
                                iq_beta = 2.50, # beta PSF of telescope + instrument
                                spaxel_size = 0.207, # spaxel size in arcsec
                                aperture = 0.93, # fiber diameter in arcsec
                                dlbda = 0.029, # Angstroem/pixel
                                lbda1 = 5067, # starting wavelength in Angstroem
                                lbda2 = 5332, # end wavelength in Angstroem
                                lsfpix = 4.5, # LSF in spectel
                                ron = 3.0, # readout noise (e-)
                                dcurrent = 3.0, # dark current (e-/pixel/h)                                
                                )
        if not skip_dataload:
            get_data(self.moshr, chan, 'moshr', CURDIR)                    

        # MOS-HR V channel 
        chan = self.moshr['channels'][2]
        self.moshr[chan] = dict(desc='WST HR spectrometer possible baseline description',  
                                version = '1.1 13/02/2023',
                                ref = '',
                                type = 'MOS',
                                iq_fwhm = 0.30, # fwhm PSF of telescope + instrument
                                iq_beta = 2.50, # beta PSF of telescope + instrument
                                spaxel_size = 0.207, # spaxel size in arcsec
                                aperture = 0.93, # fiber diameter in arcsec
                                dlbda = 0.0367, # Angstroem/pixel
                                lbda1 = 6431, # starting wavelength in Angstroem
                                lbda2 = 6768, # end wavelength in Angstroem
                                lsfpix = 4.5, # LSF in spectel
                                ron = 3.0, # readout noise (e-)
                                dcurrent = 3.0, # dark current (e-/pixel/h)                                
                                )
        if not skip_dataload:
            get_data(self.moshr, chan, 'moshr', CURDIR)                    

        # MOS-HR I channel 
        chan = self.moshr['channels'][3]
        self.moshr[chan] = dict(desc='WST HR spectrometer possible baseline description',  
                                version = '1.1 13/02/2023',
                                ref = '',
                                type = 'MOS',
                                iq_fwhm = 0.30, # fwhm PSF of telescope + instrument
                                iq_beta = 2.50, # beta PSF of telescope + instrument
                                spaxel_size = 0.207, # spaxel size in arcsec
                                aperture = 0.93, # fiber diameter in arcsec
                                dlbda = 0.0478, # Angstroem/pixel
                                lbda1 = 8380, # starting wavelength in Angstroem
                                lbda2 = 8820, # end wavelength in Angstroem
                                lsfpix = 4.5, # LSF in spectel
                                ron = 3.0, # readout noise (e-)
                                dcurrent = 3.0, # dark current (e-/pixel/h)                                
                                )
        if not skip_dataload:
            get_data(self.moshr, chan, 'moshr', CURDIR)            
        
    def info(self, ins=None, full=False):
        if ins is None:
            self._info(['ifs', 'moslr', 'moshr'], full=full)
        else:
            self._info([ins], full=full)
                
           
            

               
        
        


pyetc/vista.py
import logging
from pyetc.version import  __version__
from mpdaf.obj import Spectrum, WaveCoord
from astropy.table import Table
import os, sys
import numpy as np
from mpdaf.log import setup_logging
import astropy.units as u
from .etc import ETC


CURDIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'data/VISTA')

class VISTA(ETC):
    
    def __init__(self, log=logging.INFO):
        self.version = __version__
        self.logger = logging.getLogger(__name__)
        setup_logging(__name__, level=log, stream=sys.stdout)        
        # Telescope
        self.name = 'VISTA'
        self.tel = dict(area = np.pi*3.7**2,# effective aperture in squared meter (TBC)
                        diam = 4.0, # pirmary diameter in meter
                        )  
        # --------- MOSLR -------------
        self.moslr = {} 
        self.moslr['channels'] = ['blue','green','red']
        self.moslr['skys'] = ['darksky','greysky']        
        # MOS-LR blue channel 
        chan = self.moslr['channels'][0]
        self.moslr[chan] = dict(desc='Based on 4MOST LR throughput 5/01/2022',
                                type = 'MOS',
                                spaxel_size = 0.40, # spaxel size in arcsec TBC
                                aperture = 1.45, # fiber diameter in arcsec
                                dlbda = 0.53, # Angstroem/pixel TBC should be 0.31 ?
                                lbda1 = 3700, # starting wavelength in Angstroem
                                lbda2 = 5440, # end wavelength in Angstroem
                                lsfpix = 2.5, # LSF in spectel TBC
                                ron = 2.4, # readout noise (e-)
                                dcurrent = 1.94, # dark current (e-/pixel/h)                                
                                )
        _get_data(self.moslr, chan, 'moslr')        
        # MOS-LR green channel      
        chan = self.moslr['channels'][1] 
        self.moslr[chan] = dict(desc='Based on 4MOST LR throughput 5/01/2022',
                                type = 'MOS',
                                spaxel_size = 0.25, # spaxel size in arcsec
                                aperture = 1.605, # fiber diameter in arcsec
                                dlbda = 0.46, # Angstroem/pixel
                                lbda1 = 5250, # starting wavelength in Angstroem
                                lbda2 = 7100, # end wavelength in Angstroem
                                lsfpix = 4.5, # LSF in spectel
                                ron = 3.0, # readout noise (e-)
                                dcurrent = 3.0, # dark current (e-/pixel/h)                                
                                )     
        _get_data(self.moslr, chan, 'moslr')  
        # MOS-LR red channel      
        chan = self.moslr['channels'][2]
        self.moslr[chan] = dict(desc='Based on 4MOST LR throughput 5/01/2022',
                                type = 'MOS',
                                aperture = 1.0, # fiber diameter in arcsec
                                spaxel_size = 0.25, # spaxel size in arcsec
                                dlbda = 0.53, # Angstroem/pixel
                                lbda1 = 7000, # starting wavelength in Angstroem
                                lbda2 = 9150, # end wavelength in Angstroem
                                lsfpix = 4.5, # LSF in spectel
                                ron = 3.0, # readout noise (e-)
                                dcurrent = 3.0, # dark current (e-/pixel/h)                                
                                )        
        _get_data(self.moslr, chan, 'moslr')           
        
    def info(self):
        self.logger.info('%s ETC version: %s', self.name, self.version)
        self.logger.debug('Area: %.1f m2', self.tel['area'])      
            
        for chan in self.moslr['channels']:
            self.logger.debug('MOS-LR Channel %s', chan)
            mos = self.moslr[chan]
            self.logger.debug('\t %s', mos['desc'])
            self.logger.debug('\t Spaxel size: %.2f arcsec', mos['spaxel_size'])
            self.logger.debug('\t Fiber aperture: %.1f arcsec', mos['aperture'])
            self.logger.debug('\t Wavelength range %s A step %.2f A LSF %.1f pix', mos['instrans'].get_range(), mos['instrans'].get_step(), mos['lsfpix'])
            self.logger.debug('\t Instrument transmission peak %.2f at %.0f - min %.2f at %.0f',
                              mos['instrans'].data.max(), mos['instrans'].wave.coord(mos['instrans'].data.argmax()),
                              mos['instrans'].data.min(), mos['instrans'].wave.coord(mos['instrans'].data.argmin()))
            self.logger.debug('\t Detector RON %.1f e- Dark %.1f e-/h', mos['ron'],mos['dcurrent'])
            self.logger.debug('\t Atmospheric transmission table %s', mos['atmtrans'].filename)
            self.logger.debug('\t Dark sky emission table %s', mos['darksky'].filename)
            self.logger.debug('\t Grey sky emission table %s', mos['greysky'].filename)
            self.logger.debug('\t Instrument transmission table %s', mos['instrans'].filename)        
            
            
def _get_data(obj, chan, name): 
    ins = obj[chan]       
    ins['wave'] = WaveCoord(cdelt=ins['dlbda'], crval=ins['lbda1'], 
                            shape=int((ins['lbda2']-ins['lbda1'])/ins['dlbda']+0.5),
                            cunit=u.angstrom)
    filename = f'{name}_skytable_moon05_am1_lsfconv_{chan}.fits'
    skytab_grey = Table.read(os.path.join(CURDIR,filename))
    if abs(skytab_grey['lam'][0]*10 - ins['lbda1'])>ins['dlbda'] or \
       abs(skytab_grey['lam'][-1]*10 - ins['lbda2'])>ins['dlbda'] or \
       abs(skytab_grey['lam'][1]-skytab_grey['lam'][0])*10 - ins['dlbda']>0.01:
        raise ValueError(f'Incompatible bounds between {filename} and setup')
    ins['atmtrans'] = Spectrum(data=skytab_grey['trans'], wave=ins['wave'])
    ins['atmtrans'].filename = filename
    ins['greysky'] = Spectrum(data=skytab_grey['flux'], wave=ins['wave'])
    ins['greysky'].filename = filename
    filename = f'{name}_skytable_newmoon_am1_lsfconv_{chan}.fits'
    skytab_dark = Table.read(os.path.join(CURDIR,filename))
    if abs(skytab_dark['lam'][0]*10 - ins['lbda1'])>ins['dlbda'] or \
       abs(skytab_dark['lam'][-1]*10 - ins['lbda2'])>ins['dlbda'] or \
       abs(skytab_dark['lam'][1]-skytab_grey['lam'][0])*10 - ins['dlbda']>0.01: 
        raise ValueError(f'Incompatible bounds between {filename} and setup')  
    ins['darksky'] = Spectrum(data=skytab_dark['flux'], wave=ins['wave'])
    ins['darksky'].filename = filename
    filename = f'{name}_{chan}_noatm.fits'
    trans=Table.read(os.path.join(CURDIR,filename))
    if trans['WAVE'][0]*10 > ins['lbda1'] or \
       trans['WAVE'][-1]*10 < ins['lbda2'] :
        raise ValueError(f'Incompatible bounds between {filename} and setup')        
    ins['instrans'] = Spectrum(data=np.interp(ins['atmtrans'].wave.coord(),trans['WAVE']*10,trans['TOT']), 
                                            wave=ins['atmtrans'].wave) 
    ins['instrans'].filename = filename
